lib/kadm5: improve kadm_c_ error handling

jaltman opened the pull request:
Perform error checking for each function call and consistently return
errors at the point of failure.

Refactor functions to use a common exit path.  Preserve error messages
stored in the kadm5_client_context.context when appropriate.

Change-Id: I7aa04020e4de3454066f0d88ba805fed999dbd1a

vdukhovni commented on lib/kadm5/create_c.c:
```diff
@@ -59,35 +59,49 @@ kadm5_c_create_principal(void *server_handle,
-	return ret;
+    if (ret)
+	goto out_keep_error;
     sp = krb5_storage_from_data (&reply);
```
Ditto.

vdukhovni commented on lib/kadm5/delete_c.c:
```diff
@@ -40,7 +40,7 @@ kadm5_c_delete_principal(void *server_handle, krb5_principal princ)
     kadm5_client_context *context = server_handle;
     kadm5_ret_t ret;
-    krb5_storage *sp;
+    krb5_storage *sp = NULL;
```
DItto

vdukhovni commented on lib/kadm5/delete_c.c:
```diff
@@ -49,29 +49,39 @@ kadm5_c_delete_principal(void *server_handle, krb5_principal princ)
+	goto out;
+    ret = krb5_store_principal(sp, princ);
+    if (ret)
+	goto out;
```
Cascade...

vdukhovni commented on lib/kadm5/delete_c.c:
```diff
@@ -49,29 +49,39 @@ kadm5_c_delete_principal(void *server_handle, krb5_principal princ)
-	krb5_data_free (&reply);
-	return ENOMEM;
+	goto out_keep_error;
+    sp = krb5_storage_from_data(&reply);
```
leak.

vdukhovni commented on lib/kadm5/get_c.c:
```diff
@@ -38,47 +38,58 @@ RCSID("$Id$");
     kadm5_client_context *context = server_handle;
     kadm5_ret_t ret;
-    krb5_storage *sp;
+    krb5_storage *sp = NULL;
```
Ditto.

vdukhovni commented on lib/kadm5/get_c.c:
```diff
@@ -38,47 +38,58 @@ RCSID("$Id$");
+	goto out;
+    ret = krb5_store_int32(sp, mask);
+    if (ret)
+	goto out;
```
Ditto.

vdukhovni commented on lib/kadm5/get_c.c:
```diff
@@ -38,47 +38,58 @@ RCSID("$Id$");
-	return ret;
-    sp = krb5_storage_from_data (&reply);
+	goto out_keep_error;
+    sp = krb5_storage_from_data(&reply);
```
leak.

vdukhovni commented on lib/kadm5/get_princs_c.c:
```diff
@@ -43,50 +43,76 @@ kadm5_c_get_principals(void *server_handle,
     kadm5_client_context *context = server_handle;
     kadm5_ret_t ret;
-    krb5_storage *sp;
+    krb5_storage *sp = NULL;
```
DItto

vdukhovni commented on lib/kadm5/get_princs_c.c:
```diff
@@ -43,50 +43,76 @@ kadm5_c_get_principals(void *server_handle,
+	ret = krb5_store_string(sp, expression);
+	if (ret)
+	    goto out;
+    }
```
Cascade

vdukhovni commented on lib/kadm5/get_princs_c.c:
```diff
@@ -43,50 +43,76 @@ kadm5_c_get_principals(void *server_handle,
-	return ret;
+    if (ret)
+	goto out_keep_error;
     sp = krb5_storage_from_data (&reply);
```
leak

vdukhovni commented on lib/kadm5/get_princs_c.c:
```diff
@@ -43,50 +43,76 @@ kadm5_c_get_principals(void *server_handle,
+    *princs = calloc(tmp + 1, sizeof(**princs));
+    if (*princs == NULL) {
+	ret = ENOMEM;
+	goto out;
```
cascade

vdukhovni commented on lib/kadm5/modify_c.c:
```diff
@@ -42,40 +42,52 @@ kadm5_c_modify_principal(void *server_handle,
     kadm5_client_context *context = server_handle;
     kadm5_ret_t ret;
-    krb5_storage *sp;
+    krb5_storage *sp = NULL;
```
etc.  These should be fixed, and then I can look again.

jaltman commented:
This patch has been sitting in my dev tree for quite some time and was never quite finished.   Thanks for taking a look.

vdukhovni commented:
@jaltman will you be posting a revised update for review?

jaltman commented on lib/kadm5/chpass_c.c:
```diff
@@ -100,43 +112,60 @@ kadm5_c_chpass_principal_with_key(void *server_handle,
-	return ret;
+    if (ret)
+	goto out_keep_error;
     sp = krb5_storage_from_data (&reply);
```
fixed

jaltman commented:
I've left the `if (ret) goto out;` style because it avoids unnecessary comparisons.   I've removed the unnecessary double assignments and corrected the `krb5_storage` leaks.

nicowilliams commented on lib/kadm5/chpass_c.c:
```diff
@@ -45,7 +45,7 @@ kadm5_c_chpass_principal(void *server_handle,
     kadm5_client_context *context = server_handle;
     kadm5_ret_t ret;
-    krb5_storage *sp;
+    krb5_storage *sp = NULL;
```
I agree, but sometimes VC forces one to do this sort of thing, and we in Heimdal have been victimized by VC a few times, so I understand how unnecessary initialization ends up becoming second nature.

nicowilliams commented on lib/kadm5/chpass_c.c:
```diff
@@ -59,36 +59,49 @@ kadm5_c_chpass_principal(void *server_handle,
     if (sp == NULL) {
-	krb5_clear_error_message(context->context);
-	return ENOMEM;
+	ret = ENOMEM;
```
`ret = krb5_enomem(context);` would be better -- that's what `krb5_enomem()` exists for.

nicowilliams commented on lib/kadm5/chpass_c.c:
```diff
@@ -62,33 +62,45 @@ kadm5_c_chpass_principal(void *server_handle,
+	goto out;
+    ret = krb5_store_int32(sp, keepold); /* extension */
+    if (ret)
+	goto out;
```
How?

jaltman replied:
On failure the cascade style requires that `if (ret == 0)` be compared for each statement that will be skipped.

nicowilliams replied:
That's not an optimization we need though.

jaltman replied:
@nicowilliams As I started to convert to your preferred style it ends up being problematic because some of the errors we want to return after clearing the krb5_context error state and others we do not.   

I'm updating the patch for the krb5_enomem() changes but otherwise leaving it as-is.

nicowilliams replied:
Ok.

nicowilliams commented on lib/kadm5/chpass_c.c:
```diff
@@ -59,36 +59,49 @@ kadm5_c_chpass_principal(void *server_handle,
-	krb5_clear_error_message(context->context);
-	krb5_data_free (&reply);
-	return ENOMEM;
+	ret = ENOMEM;
```
Another `krb5_enomem()` opportunity.

jaltman commented:
@nicowilliams I will push a separate commit with the krb5_enomem() change since that is broader than just the locations I modified.

lhoward commented on lib/kadm5/create_c.c:
```diff
@@ -59,35 +59,50 @@ kadm5_c_create_principal(void *server_handle,
-    krb5_ret_int32(sp, &tmp);
+    ret = krb5_ret_int32(sp, &tmp);
+    if (ret == 0)
+	ret = tmp;;
```
double semicolon

jaltman replied:
fixed

lhoward commented on lib/kadm5/get_princs_c.c:
```diff
@@ -47,46 +47,73 @@ kadm5_c_get_principals(void *server_handle,
+    if (ret == 0)
+	ret = tmp;
+
+    if (ret == 0)
```
should this be if ret !=0 ?

jaltman replied:
fixed

nicowilliams commented:
Thanks for the `krb5_enomem()` changes.

jaltman commented:
I only updated the lib/kadm5 directory.  The same conversion cod be performed one directory at a time.
